### PR TITLE
Generate endpoint providers

### PR DESCRIPTION
**Pull request by dagnir**
## Motivation and Context
Further support for rules-based endpoint resolution.

## Modifications
This commit adds support for generating endpoint provider implementations based
on the given endpoint rule sets. The generated providers currently delegate to
the runtime rules engine for evaluation.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/IntermediateModelBuilder.java [hunk 1]
```diff
@@ -68,6 +70,7 @@ public IntermediateModelBuilder(C2jModels models) {
         this.shapeProcessors = createShapeProcessors();
         this.paginators = models.paginatorsModel();
         this.waiters = models.waitersModel();
+        this.endpointRuleSet = models.endpointRuleSetModel();
```
[the remark this answers is not in the document]
Comment (dagnir) [reply codegen/src/main/java/software/amazon/awssdk/codegen/IntermediateModelBuilder.java:73]: They're different, `EndpointRuleSet` is the actual structure intepreted by the `RuleEngine`. The model is the codegen's view of it

File: codegen/src/main/java/software/amazon/awssdk/codegen/model/service/EndpointRuleSetModel.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    public Map<String, ParameterModel> getParameters() {
+        return parameters;
+    }
+
+    public void setParameters(Map<String, ParameterModel> parameters) {
+        this.parameters = parameters;
+    }
+
+    public List<JsonNode> getRules() {
+        return rules;
```
[the remark this answers is not in the document]
Comment (dagnir) [reply codegen/src/main/java/software/amazon/awssdk/codegen/model/service/EndpointRuleSetModel.java:53]: I think in this case it's fine, it's just an internal class to the code generator

File: codegen/src/main/java/software/amazon/awssdk/codegen/internal/Constant.java [hunk 3]
```diff
@@ -68,6 +68,8 @@ public final class Constant {
 
     public static final String PACKAGE_NAME_WAITERS_PATTERN = "%s.waiters";
 
+    public static final String PACKAGE_NAME_RULES_PATTERN = "%s.rules";
```
[the remark this answers is not in the document]
Comment (dagnir) [reply codegen/src/main/java/software/amazon/awssdk/codegen/internal/Constant.java:71]: I think we ironed out some of these during the last team review. I'll just keep this change here and make all of the package changes in one place.

File: codegen/src/main/java/software/amazon/awssdk/codegen/model/service/ParameterModel.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.model.service;
+
+public class ParameterModel {
+    private String type;
+
+    public String getType() {
```
[the remark this answers is not in the document]
Comment (dagnir) [reply codegen/src/main/java/software/amazon/awssdk/codegen/model/service/ParameterModel.java:21]: Done

**Review by joviegas — APPROVED** [review]

**Pull request merged**